### ISSUE TITLE
non-finite loss, ending training

### ISSUE THREAD

**Issue by melonux** [ultralytics/yolov5#673]
## ❔Question
repo: tag v2.0

cmd: python -u train.py --img-size 800 --device 0 --batch 16 --epochs 1000 --data ./pyq6/dataset.yaml --cfg ./pyq6/yolov5.yaml --weight weights/yolov5s.pt

problem: If fl_gamma is *not* equal to 0.0, then train on my dataset with or without [--rect] will get "non-finite loss, ending training" error in 10-20 epochs. It is repeatable. I see related discussions in yolov3, but yolov5 seems to have the same problem.

## Additional context
```
Hyperparameters {'optimizer': 'SGD', 'lr0': 0.01, 'momentum': 0.937, 'weight_decay': 0.0005, 'giou': 0.05, 'cls': 0.5, 'cls_pw': 1.0, 'obj': 1.0, 'obj_pw': 1.0, 'iou_t': 0.2, 'anchor_t': 4.0, 'fl_gamma': 0.5, 'hsv_h': 0.015, 'hsv_s': 0.7, 'hsv_v': 0.4, 'degrees': 0.0, 'translate': 0.0, 'scale': 0.2, 'shear': 0.0}
 
                from  n    params  module                                  arguments                     

  0                -1  1      3520  models.common.Focus                     [3, 32, 3]                    
  1                -1  1     18560  models.common.Conv                      [32, 64, 3, 2]                
  2                -1  1     19904  models.common.BottleneckCSP             [64, 64, 1]                   
  3                -1  1     73984  models.common.Conv                      [64, 128, 3, 2]               
  4                -1  1    161152  models.common.BottleneckCSP             [128, 128, 3]                 
  5                -1  1    295424  models.common.Conv                      [128, 256, 3, 2]              
  6                -1  1    641792  models.common.BottleneckCSP             [256, 256, 3]                 
  7                -1  1   1180672  models.common.Conv                      [256, 512, 3, 2]              
  8                -1  1    656896  models.common.SPP                       [512, 512, [5, 9, 13]]        
  9                -1  1   1248768  models.common.BottleneckCSP             [512, 512, 1, False]          
 10                -1  1    131584  models.common.Conv                      [512, 256, 1, 1]              
 11                -1  1         0  torch.nn.modules.upsampling.Upsample    [None, 2, 'nearest']          
 12           [-1, 6]  1         0  models.common.Concat                    [1]                           
 13                -1  1    378624  models.common.BottleneckCSP             [512, 256, 1, False]          
 14                -1  1     33024  models.common.Conv                      [256, 128, 1, 1]              
 15                -1  1         0  torch.nn.modules.upsampling.Upsample    [None, 2, 'nearest']          
 16           [-1, 4]  1         0  models.common.Concat                    [1]                           
 17                -1  1     95104  models.common.BottleneckCSP             [256, 128, 1, False]          
 18                -1  1    147712  models.common.Conv                      [128, 128, 3, 2]              
 19          [-1, 14]  1         0  models.common.Concat                    [1]                           
 20                -1  1    313088  models.common.BottleneckCSP             [256, 256, 1, False]          
 21                -1  1    590336  models.common.Conv                      [256, 256, 3, 2]              
 22          [-1, 10]  1         0  models.common.Concat                    [1]                           
 23                -1  1   1248768  models.common.BottleneckCSP             [512, 512, 1, False]          
 24      [17, 20, 23]  1     83607  models.yolo.Detect                      [26, [[10, 13, 16, 30, 33, 23], [30, 61, 62, 45, 59, 119], [116, 90, 156, 198, 373, 326]], [128, 256, 512]]
Model Summary: 191 layers, 7.32252e+06 parameters, 7.32252e+06 gradients

Optimizer groups: 62 .bias, 70 conv.weight, 59 other
Transferred 362/370 items from weights/yolov5s.pt
Scanning images: 100%|██████████| 424/424 [00:00<00:00, 5706.50it/s]
Scanning labels pyq6/labels.cache (335 found, 0 missing, 89 empty, 0 duplicate, for 424 images): 100%|██████████| 424/424 [00:00<00:00, 14614.06it/s]
Scanning images: 100%|██████████| 10/10 [00:00<00:00, 5260.63it/s]
Scanning labels pyq6/labels.cache (9 found, 0 missing, 1 empty, 0 duplicate, for 10 images): 100%|██████████| 10/10 [00:00<00:00, 9927.35it/s]

Analyzing anchors... Best Possible Recall (BPR) = 0.9997
Image sizes 800 train, 800 test
Using 6 dataloader workers
Starting training for 1000 epochs...

     Epoch   gpu_mem      GIoU       obj       cls     total   targets  img_size
     0/999     5.57G    0.1296  0.008727   0.02216    0.1604       158       800: 100%|██████████| 27/27 [00:26<00:00,  1.02it/s]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100%|██████████| 1/1 [00:03<00:00,  3.58s/it]
                 all          10           0           0           0           0           0

     Epoch   gpu_mem      GIoU       obj       cls     total   targets  img_size
     1/999     7.37G    0.1098   0.01316   0.02047    0.1434       189       800: 100%|██████████| 27/27 [00:18<00:00,  1.44it/s]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100%|██████████| 1/1 [00:01<00:00,  1.61s/it]
                 all          10          43           0           0    0.000373    0.000187

     Epoch   gpu_mem      GIoU       obj       cls     total   targets  img_size
     2/999     7.37G   0.09402   0.01581   0.01904    0.1289       110       800: 100%|██████████| 27/27 [00:18<00:00,  1.48it/s]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100%|██████████| 1/1 [00:01<00:00,  1.48s/it]
                 all          10          43           0           0    0.000845    0.000426

     Epoch   gpu_mem      GIoU       obj       cls     total   targets  img_size
     3/999     7.38G   0.07822   0.02007   0.01814    0.1164       168       800: 100%|██████████| 27/27 [00:18<00:00,  1.47it/s]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100%|██████████| 1/1 [00:01<00:00,  1.50s/it]
                 all          10          43           0           0     0.00534     0.00127

     Epoch   gpu_mem      GIoU       obj       cls     total   targets  img_size
     4/999     7.38G   0.06358   0.02069   0.01755    0.1018       150       800: 100%|██████████| 27/27 [00:18<00:00,  1.45it/s]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100%|██████████| 1/1 [00:01<00:00,  1.75s/it]
                 all          10          43           0           0      0.0125     0.00423

     Epoch   gpu_mem      GIoU       obj       cls     total   targets  img_size
     5/999     7.38G   0.05436   0.01992   0.01681   0.09108        96       800: 100%|██████████| 27/27 [00:18<00:00,  1.47it/s]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100%|██████████| 1/1 [00:02<00:00,  2.40s/it]
                 all          10          43       0.111       0.025      0.0675      0.0389

     Epoch   gpu_mem      GIoU       obj       cls     total   targets  img_size
     6/999     7.38G   0.05421   0.01755   0.01619   0.08796        71       800: 100%|██████████| 27/27 [00:17<00:00,  1.50it/s]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100%|██████████| 1/1 [00:02<00:00,  2.51s/it]
                 all          10          43      0.0556      0.0111      0.0819      0.0398

     Epoch   gpu_mem      GIoU       obj       cls     total   targets  img_size
     7/999     7.38G   0.04873   0.01611   0.01548   0.08033       137       800: 100%|██████████| 27/27 [00:17<00:00,  1.50it/s]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100%|██████████| 1/1 [00:01<00:00,  1.99s/it]
                 all          10          43       0.201      0.0746       0.186       0.089

     Epoch   gpu_mem      GIoU       obj       cls     total   targets  img_size
     8/999     7.38G   0.04707   0.01446   0.01494   0.07647       118       800: 100%|██████████| 27/27 [00:17<00:00,  1.50it/s]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100%|██████████| 1/1 [00:01<00:00,  1.73s/it]
                 all          10          43       0.293       0.136       0.264       0.153

     Epoch   gpu_mem      GIoU       obj       cls     total   targets  img_size
     9/999     7.38G    0.0451   0.01478   0.01404   0.07392       117       800: 100%|██████████| 27/27 [00:17<00:00,  1.51it/s]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100%|██████████| 1/1 [00:01<00:00,  1.68s/it]
                 all          10          43        0.18       0.192       0.277       0.129

     Epoch   gpu_mem      GIoU       obj       cls     total   targets  img_size
    10/999     7.38G   0.04287   0.01377   0.01335   0.06998       139       800: 100%|██████████| 27/27 [00:18<00:00,  1.50it/s]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100%|██████████| 1/1 [00:01<00:00,  1.55s/it]
                 all          10          43       0.358       0.242       0.369       0.236

     Epoch   gpu_mem      GIoU       obj       cls     total   targets  img_size
    11/999     7.38G   0.04122    0.0134   0.01295   0.06757       190       800: 100%|██████████| 27/27 [00:18<00:00,  1.50it/s]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100%|██████████| 1/1 [00:01<00:00,  1.58s/it]
                 all          10          43       0.315       0.286       0.435       0.293

     Epoch   gpu_mem      GIoU       obj       cls     total   targets  img_size
    12/999     7.38G    0.0387   0.01212   0.01257   0.06339       149       800: 100%|██████████| 27/27 [00:18<00:00,  1.49it/s]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100%|██████████| 1/1 [00:01<00:00,  1.61s/it]
                 all          10          43       0.275        0.35       0.441       0.321

     Epoch   gpu_mem      GIoU       obj       cls     total   targets  img_size
    13/999     7.38G   0.03785    0.0118   0.01222   0.06187       132       800: 100%|██████████| 27/27 [00:18<00:00,  1.47it/s]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100%|██████████| 1/1 [00:01<00:00,  1.54s/it]
                 all          10          43       0.345       0.353       0.452       0.281

     Epoch   gpu_mem      GIoU       obj       cls     total   targets  img_size
    14/999     7.38G    0.0373   0.01103   0.01193   0.06027        78       800: 100%|██████████| 27/27 [00:18<00:00,  1.49it/s]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100%|██████████| 1/1 [00:01<00:00,  1.44s/it]
                 all          10          43         0.3        0.31       0.429       0.297

     Epoch   gpu_mem      GIoU       obj       cls     total   targets  img_size
    15/999     7.38G   0.03664   0.01046   0.01146   0.05855       139       800: 100%|██████████| 27/27 [00:17<00:00,  1.50it/s]
               Class      Images     Targets           P           R      mAP@.5  mAP@.5:.95: 100%|██████████| 1/1 [00:01<00:00,  1.47s/it]
                 all          10          43       0.405        0.41       0.548       0.362

     Epoch   gpu_mem      GIoU       obj       cls     total   targets  img_size
    16/999     7.38G   0.03433   0.01034   0.01147   0.05613       130       800:  85%|████████▌ | 23/27 [00:15<00:02,  1.66it/s]WARNING: non-finite loss, ending training  tensor([nan, nan, nan, nan], device='cuda:0')
    16/999     7.38G   0.03433   0.01034   0.01147   0.05613       130       800:  85%|████████▌ | 23/27 [00:16<00:02,  1.43it/s]```

**Comment by glenn-jocher** [comment]
@melonux your code is out of date. Suggest you git clone the latest.
